riscv_analysis/src/parser/lexer.rs

```diff
@@ -759,4 +831,98 @@ mod tests {
             )
         ));
     }
+
+    #[test]
+    fn chars() {
+        let input = "'a' 'b' '\\'' '\\u03bb'";
+        let tokens = tokenize(input);
+
+        assert_eq!(
+            tokens,
+            vec![
+                Token::Char('a'),
+                Token::Char('b'),
+                Token::Char('\''),
+                Token::Char('\u{03bb}'),
+            ]
+        );
+    }
+
+    #[test]
+    fn unclosed_char() {
+        let input = "'a \n 'b";
+        let tokens = tokenize_err(input);
+        assert_eq!(tokens.len(), 3);
+
+        assert!(matches!(
+            &tokens[0],
+            Err(err) if matches!(
+                err,
+                LexError::InvalidString(_info, sub)
+                    if sub.kind == StringLexErrorType::Unclosed
+            )
+        ));
+
+        assert!(matches!(
+            &tokens[1],
+            Ok(info) if matches!(&info.token, Token::Newline)
+        ));
+
+        assert!(matches!(
+            &tokens[2],
+            Err(err) if matches!(
+                err,
+                LexError::InvalidString(_info, sub)
+                    if sub.kind == StringLexErrorType::Unclosed
+            )
+        ));
+    }
+
+    #[test]
+    fn newline_in_char() {
+        let input = "'\n'";
+        let tokens = tokenize_err(input);
+        assert_eq!(tokens.len(), 3);
+
+        assert!(matches!(
+            &tokens[0],
+            Err(err) if matches!(
+                err,
+                LexError::InvalidString(_info, sub)
+                    if sub.kind == StringLexErrorType::Newline
+            )
+        ));
+
+        assert!(matches!(
+            &tokens[1],
+            Ok(info) if matches!(&info.token, Token::Newline)
+        ));
+
+        assert!(matches!(
+            &tokens[2],
+            Err(err) if matches!(
+                err,
+                LexError::InvalidString(_info, sub)
+                    if sub.kind == StringLexErrorType::Unclosed
+            )
+        ));
+    }
+
```

Comment:
I would add this newline test case. Your code passes it, but I thought it would be good to include.

```suggestion

    #[test]
    fn newline_chars() {
        let input = "'\\n'  'n'  \n  ";
        let tokens = tokenize(input);

        assert_eq!(
            tokens,
            vec![Token::Char('\n'), Token::Char('n'), Token::Newline]
        );
    }
    
```